File: frontend/src/api/schedule.ts
import { apiClient } from './client'

export interface ShiftTemplate {
  id: string
  org_id: string
  name: string
  start_time: string
  end_time: string
  crosses_midnight: boolean
  duration_minutes: number
  color: string
  is_active: boolean
  created_at: string
  /** Present when backend returns it; used for optimistic locking. */
  updated_at?: string
}

export interface AssignmentView {
  assignment_id: string
  date: string
  shift_name: string
  shift_color: string
  start_time: string
  end_time: string
  crosses_midnight: boolean
  user_id: string
  employee_id: string | null
  first_name: string
  last_name: string
  position: string | null
  is_overtime: boolean
  is_trade: boolean
  team_name: string | null
  classification_abbreviation: string | null
  notes: string | null
}

// -- Grid / Day / Dashboard types --

export interface GridAssignment {
  assignment_id: string
  user_id: string
  employee_id: string | null
  first_name: string
  last_name: string
  classification_abbreviation: string | null
  is_overtime: boolean
  is_trade: boolean
  notes?: string | null
}

export interface ClassificationCoverageDetail {
  classification_abbreviation: string
  shortage: number
}

export interface GridCell {
  date: string
  shift_template_id: string
  shift_name: string
  shift_color: string
  assignments: GridAssignment[]
  leave_count: number
  coverage_required: number
  coverage_actual: number
  coverage_by_classification?: ClassificationCoverageDetail[]
}

export interface DayViewEntry {
  shift_template_id: string
  shift_name: string
  shift_color: string
  start_time: string
  end_time: string
  crosses_midnight: boolean
  assignments: GridAssignment[]
  coverage_required: number
  coverage_actual: number
  coverage_status: 'green' | 'yellow' | 'red'
  coverage_by_classification?: ClassificationCoverageDetail[]
}

export interface ScheduleAnnotation {
  id: string
  org_id: string
  date: string
  shift_template_id: string | null
  content: string
  annotation_type: 'note' | 'alert' | 'holiday'
  created_by: string
  created_at: string
}

export interface DashboardData {
  current_coverage: DayViewEntry[]
  pending_leave_count: number
  open_callout_count: number
  annotations: ScheduleAnnotation[]
}

export const scheduleApi = {
  getTemplates: () =>
    apiClient.get<ShiftTemplate[]>('/api/shifts/templates'),

  createTemplate: (body: {
    name: string
    start_time: string
    end_time: string
    color?: string
  }) => apiClient.post<ShiftTemplate>('/api/shifts/templates', body),

  updateTemplate: (id: string, body: { name?: string; color?: string; is_active?: boolean; expected_updated_at?: string }) =>
    apiClient.patch<ShiftTemplate>(`/api/shifts/templates/${id}`, body),

  getStaffing: (start_date: string, end_date: string, team_id?: string) =>
    apiClient.get<AssignmentView[]>('/api/schedule', {
      params: { start_date, end_date, ...(team_id ? { team_id } : {}) },
    }),

  createAssignment: (body: {
    scheduled_shift_id: string
    user_id: string
    position?: string
    is_overtime?: boolean
    is_trade?: boolean
    notes?: string
  }) => apiClient.post('/api/schedule/assignments', body),

  deleteAssignment: (id: string) =>
    apiClient.delete(`/api/schedule/assignments/${id}`),

  // Grid view
  getGrid: (start_date: string, end_date: string, team_id?: string) =>
    apiClient.get<GridCell[]>('/api/schedule/grid', {
      params: { start_date, end_date, ...(team_id ? { team_id } : {}) },
    }),

  // Day view
  getDayView: (date: string) =>
    apiClient.get<DayViewEntry[]>(`/api/schedule/day/${date}`),

  // Dashboard
  getDashboard: () =>
    apiClient.get<DashboardData>('/api/schedule/dashboard'),

  // Annotations
  listAnnotations: (start_date: string, end_date: string) =>
    apiClient.get<ScheduleAnnotation[]>('/api/schedule/annotations', {
      params: { start_date, end_date },
    }),

  createAnnotation: (body: {
    date: string
    shift_template_id?: string
    content: string
    annotation_type: string
  }) => apiClient.post<ScheduleAnnotation>('/api/schedule/annotations', body),

  deleteAnnotation: (id: string) =>
    apiClient.delete(`/api/schedule/annotations/${id}`),
}

File: backend/src/models/schedule.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use uuid::Uuid;
use validator::Validate;

/// An assignment of a specific employee to a specific scheduled shift.
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Assignment {
    pub id: Uuid,
    pub scheduled_shift_id: Uuid,
    pub user_id: Uuid,
    pub position: Option<String>,
    pub is_overtime: bool,
    pub is_trade: bool,
    pub notes: Option<String>,
    pub created_by: Uuid,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

/// Rich view joining assignment + shift + user for the staffing board.
#[derive(Debug, Clone, Serialize)]
pub struct AssignmentView {
    pub assignment_id: Uuid,
    pub date: time::Date,
    pub shift_name: String,
    pub shift_color: String,
    #[serde(with = "crate::models::common::time_format")]
    pub start_time: time::Time,
    #[serde(with = "crate::models::common::time_format")]
    pub end_time: time::Time,
    pub crosses_midnight: bool,
    pub user_id: Uuid,
    pub employee_id: Option<String>,
    pub first_name: String,
    pub last_name: String,
    pub position: Option<String>,
    pub is_overtime: bool,
    pub is_trade: bool,
    pub team_name: Option<String>,
    pub classification_abbreviation: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateAssignmentRequest {
    pub scheduled_shift_id: Uuid,
    pub user_id: Uuid,
    #[validate(length(max = 100))]
    pub position: Option<String>,
    #[serde(default)]
    pub is_overtime: bool,
    #[serde(default)]
    pub is_trade: bool,
    #[validate(length(max = 2000))]
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct StaffingQuery {
    pub start_date: time::Date,
    pub end_date: time::Date,
    pub team_id: Option<Uuid>,
}

// -- Coverage Plans (per-half-hour-slot system) --

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CoveragePlan {
    pub id: Uuid,
    pub org_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub is_default: bool,
    pub is_active: bool,
    pub created_by: Uuid,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
}

#[derive(Debug, Serialize)]
pub struct CoveragePlanView {
    pub id: Uuid,
    pub org_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub is_default: bool,
    pub is_active: bool,
    pub slot_count: i64,
    pub assignment_count: i64,
    pub created_by: Uuid,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CoveragePlanSlot {
    pub id: Uuid,
    pub plan_id: Uuid,
    pub classification_id: Uuid,
    pub day_of_week: i16,
    pub slot_index: i16,
    pub min_headcount: i16,
    pub target_headcount: i16,
    pub max_headcount: i16,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CoveragePlanAssignment {
    pub id: Uuid,
    pub org_id: Uuid,
    pub plan_id: Uuid,
    pub start_date: time::Date,
    pub end_date: Option<time::Date>,
    pub notes: Option<String>,
    pub created_by: Uuid,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateCoveragePlanRequest {
    #[validate(length(min = 1, max = 100))]
    pub name: String,
    #[validate(length(max = 500))]
    pub description: Option<String>,
    #[serde(default)]
    pub is_default: bool,
}

#[derive(Debug, Deserialize)]
pub struct UpdateCoveragePlanRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub is_default: Option<bool>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct SlotEntry {
    pub classification_id: Uuid,
    pub day_of_week: i16,
    pub slot_index: i16,
    pub min_headcount: i16,
    pub target_headcount: i16,
    pub max_headcount: i16,
}

#[derive(Debug, Deserialize)]
pub struct BulkUpsertSlotsRequest {
    pub slots: Vec<SlotEntry>,
}

#[derive(Debug, Deserialize)]
pub struct CreateCoveragePlanAssignmentRequest {
    pub plan_id: Uuid,
    pub start_date: time::Date,
    pub end_date: Option<time::Date>,
    pub notes: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SlotCoverage {
    pub slot_index: i16,
    pub classification_id: Uuid,
    pub classification_abbreviation: String,
    pub min_headcount: i16,
    pub target_headcount: i16,
    pub max_headcount: i16,
    pub actual_headcount: i32,
    pub status: String,
}

// -- Classification Gap (per-shift coverage shortage) --

#[derive(Debug, Serialize)]
pub struct ClassificationGap {
    pub classification_id: Uuid,
    pub classification_abbreviation: String,
    pub shift_template_id: Uuid,
    pub shift_name: String,
    pub shift_color: String,
    pub target: i16,
    pub actual: i16,
    pub shortage: i16,
}

// -- Block-level coverage gaps (contiguous time ranges per classification) --

#[derive(Debug, Clone, Serialize)]
pub struct CoverageGapBlock {
    pub start_time: String,
    pub end_time: String,
    pub shortage: i32,
}

#[derive(Debug, Serialize)]
pub struct ClassificationGapBlocks {
    pub classification_id: Uuid,
    pub classification_abbreviation: String,
    pub blocks: Vec<CoverageGapBlock>,
}

// -- Schedule Annotations --

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ScheduleAnnotation {
    pub id: Uuid,
    pub org_id: Uuid,
    pub date: time::Date,
    pub shift_template_id: Option<Uuid>,
    pub content: String,
    pub annotation_type: String,
    pub created_by: Uuid,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateAnnotationRequest {
    pub date: time::Date,
    pub shift_template_id: Option<Uuid>,
    #[validate(length(max = 5000))]
    pub content: String,
    #[validate(length(max = 50))]
    pub annotation_type: String,
}

#[derive(Debug, Deserialize)]
pub struct AnnotationQuery {
    pub start_date: time::Date,
    pub end_date: time::Date,
}

// -- Per-classification coverage detail (included in API responses) --

#[derive(Debug, Clone, Serialize)]
pub struct ClassificationCoverageDetail {
    pub classification_abbreviation: String,
    pub shortage: i32,
}

// -- Grid View Types --

#[derive(Debug, Serialize)]
pub struct GridCell {
    pub date: time::Date,
    pub shift_template_id: Uuid,
    pub shift_name: String,
    pub shift_color: String,
    pub assignments: Vec<GridAssignment>,
    pub leave_count: i64,
    pub coverage_required: i32,
    pub coverage_actual: i32,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub coverage_by_classification: Vec<ClassificationCoverageDetail>,
}

#[derive(Debug, Serialize)]
pub struct GridAssignment {
    pub assignment_id: Uuid,
    pub user_id: Uuid,
    pub employee_id: Option<String>,
    pub first_name: String,
    pub last_name: String,
    pub classification_abbreviation: Option<String>,
    pub is_overtime: bool,
    pub is_trade: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
}

// -- Day View Types --

#[derive(Debug, Serialize)]
pub struct DayViewEntry {
    pub shift_template_id: Uuid,
    pub shift_name: String,
    pub shift_color: String,
    #[serde(with = "crate::models::common::time_format")]
    pub start_time: time::Time,
    #[serde(with = "crate::models::common::time_format")]
    pub end_time: time::Time,
    pub crosses_midnight: bool,
    pub assignments: Vec<GridAssignment>,
    pub coverage_required: i32,
    pub coverage_actual: i32,
    pub coverage_status: String,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub coverage_by_classification: Vec<ClassificationCoverageDetail>,
}

// -- Day Grid Types (coverage day view with employee-level detail) --

#[derive(Debug, Serialize)]
pub struct DayGridResponse {
    pub date: String,
    pub classifications: Vec<DayGridClassification>,
    pub blocks: Vec<CoverageBlock>,
}

#[derive(Debug, Serialize)]
pub struct DayGridClassification {
    pub classification_id: Uuid,
    pub abbreviation: String,
    pub blocks: Vec<ClassificationBlock>,
}

#[derive(Debug, Serialize)]
pub struct ClassificationBlock {
    pub block_index: u8,
    pub start_time: String,
    pub end_time: String,
    pub min: i16,
    pub target: i16,
    pub actual: i32,
    pub status: String,
    pub employees: Vec<BlockEmployee>,
}

#[derive(Debug, Clone, Serialize)]
pub struct BlockEmployee {
    pub user_id: Uuid,
    pub first_name: String,
    pub last_name: String,
    pub shift_name: String,
    pub shift_start: String,
    pub shift_end: String,
    pub is_overtime: bool,
    pub assignment_id: Uuid,
}

#[derive(Debug, Serialize)]
pub struct CoverageBlock {
    pub block_index: u8,
    pub total_target: i32,
    pub total_actual: i32,
    pub status: String,
}

// -- Dashboard Types --

#[derive(Debug, Serialize)]
pub struct DashboardData {
    pub current_coverage: Vec<DayViewEntry>,
    pub pending_leave_count: i64,
    pub open_callout_count: i64,
    pub annotations: Vec<ScheduleAnnotation>,
}
